Concurrent execution, SimpleDateFormat is not threadsafe

I hit into this when having concurrent execution. Looks like the SimpleDateFormat is causing the problem.

https://github.com/introproventures/graphql-jpa-query/blob/master/graphql-jpa-query-schema/src/main/java/com/introproventures/graphql/jpa/query/schema/JavaScalars.java
public GraphQLDateCoercing() {
            df = new SimpleDateFormat("yyyy-MM-dd");
}

private Date parseStringToDate(String input) {
            try {
                return df.parse(input);
            } catch (ParseException e) {
                log.warn("Failed to parse Date from input: " + input, e);
                return null;
            }
}


> java.util.concurrent.CompletionException: java.lang.NumberFormatException: multiple points
>                 at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292)
>                 at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308)
>                 at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:769)
>                 at java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:778)
>                 at java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2140)
>                 at graphql.execution.Execution.executeOperation(Execution.java:174)
>                 at graphql.execution.Execution.execute(Execution.java:101)
>                 at graphql.GraphQL.execute(GraphQL.java:573)
>                 at graphql.GraphQL.parseValidateAndExecute(GraphQL.java:515)
>                 at graphql.GraphQL.executeAsync(GraphQL.java:489)
>                 at graphql.GraphQL.execute(GraphQL.java:420)
>                 at graphql.GraphQL.execute(GraphQL.java:390)
>                 at com.introproventures.graphql.jpa.query.schema.impl.GraphQLJpaExecutor.execute(GraphQLJpaExecutor.java:101)
>                 at com.introproventures.graphql.jpa.query.schema.impl.GraphQLJpaExecutor$$FastClassBySpringCGLIB$$3c8ebb24.invoke(<generated>)
>                 at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
> org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:836)
>                 at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1747)
> Caused by: java.lang.NumberFormatException: multiple points
>                 at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:1890)
>                 at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
>                 at java.lang.Double.parseDouble(Double.java:538)
>                 at java.text.DigitList.getDouble(DigitList.java:169)
>                 at java.text.DecimalFormat.parse(DecimalFormat.java:2089)
>                 at java.text.SimpleDateFormat.subParse(SimpleDateFormat.java:2162)
>                 at java.text.SimpleDateFormat.parse(SimpleDateFormat.java:1514)
>                 at java.text.DateFormat.parse(DateFormat.java:364)
>                 at com.introproventures.graphql.jpa.query.schema.JavaScalars$GraphQLDateCoercing.parseStringToDate(JavaScalars.java:335)
>                 at com.introproventures.graphql.jpa.query.schema.JavaScalars$GraphQLDateCoercing.parseLiteral(JavaScalars.java:325)
>                 at graphql.schema.Coercing.parseLiteral(Coercing.java:89)
>   graphql.execution.ValuesResolver.coerceValueAstForInputObject(ValuesResolver.java:306)
>                 at graphql.execution.ValuesResolver.coerceValueAst(ValuesResolver.java:255)
>                 at graphql.execution.ValuesResolver.coerceValueAstForInputObject(ValuesResolver.java:306)
>                 at graphql.execution.ValuesResolver.coerceValueAst(ValuesResolver.java:255)
> graphql.execution.ExecutionStrategy.lambda$resolveFieldWithInfo$0(ExecutionStrategy.java:204)
>                 at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
>  
>

--- a/graphql-jpa-query-schema/src/main/java/com/introproventures/graphql/jpa/query/schema/JavaScalars.java
+++ b/graphql-jpa-query-schema/src/main/java/com/introproventures/graphql/jpa/query/schema/JavaScalars.java
@@ -290,14 +290,14 @@ public class JavaScalars {
     }
 
     public static class GraphQLDateCoercing implements Coercing<Object, Object> {
-        final DateFormat df;
+        final String dateFormatString;
 
 
         /**
          * Default to pattern 'yyyy-MM-dd'
          */
         public GraphQLDateCoercing() {
-            df = new SimpleDateFormat("yyyy-MM-dd");
+        	dateFormatString = "yyyy-MM-dd";
         }
 
         /**
@@ -306,7 +306,7 @@ public class JavaScalars {
          * @param dateFormatString e.g. "yyyy-MM-dd'T'HH:mm:ss.SSSXXX" for "2001-07-04T12:08:56.235-07:00"
          */
         public GraphQLDateCoercing(String dateFormatString) {
-            df = new SimpleDateFormat(dateFormatString);
+        	this.dateFormatString = dateFormatString;
         }
 
         @Override
@@ -340,6 +340,8 @@ public class JavaScalars {
         }
 
         private Date parseStringToDate(String input) {
+        	DateFormat df = new SimpleDateFormat(dateFormatString);
+        	
             try {
                 return df.parse(input);
             } catch (ParseException e) {
@@ -662,7 +664,7 @@ public class JavaScalars {
 
         @Override
         public Object parseLiteral(Object input) {
-            return parseLiteral((Value<?>) input, Collections.emptyMap());
+            return parseLiteral(input, Collections.emptyMap());
         }
 
         //recursively parse the input into a Map

--- a/graphql-jpa-query-schema/src/test/java/com/introproventures/graphql/jpa/query/schema/JavaScalarsTest.java
+++ b/graphql-jpa-query-schema/src/test/java/com/introproventures/graphql/jpa/query/schema/JavaScalarsTest.java
@@ -28,7 +28,11 @@ import java.time.OffsetDateTime;
 import java.time.ZoneId;
 import java.time.ZoneOffset;
 import java.time.ZonedDateTime;
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 
 import org.junit.Test;
@@ -289,4 +293,29 @@ public class JavaScalarsTest {
         assert resultLDT.getMinute() == 15;
         assert resultLDT.getSecond() == 07;
     }
+    
+    @Test
+    public void dateCoercionThreadSafe() throws InterruptedException, ExecutionException {
+    	//given
+    	String dateLiteral = "2018-06-22T10:00:00";
+        Coercing<?, ?> subject = new JavaScalars.GraphQLDateCoercing();
+    	
+        List<CompletableFuture<Object>> dates = new ArrayList<>();
+
+        //when
+        for(int i = 0; i < 1000; i++) {
+            CompletableFuture<Object> task = CompletableFuture.supplyAsync(() -> {
+            	return subject.serialize(dateLiteral);
+            });
+        	dates.add(task);
+        }
+        
+        CompletableFuture<Void> result = CompletableFuture.allOf(dates.toArray(new CompletableFuture[] {}));
+        
+        result.join();
+        
+        //then
+        assertThat(result.isCompletedExceptionally()).isFalse();        
+    }
+    
 }

